lib/Dialect/RTL/RTLFolds.cpp

```diff
@@ -706,3 +714,259 @@ void ConcatOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
   };
   results.insert<Folder>(context);
 }
+
+//===----------------------------------------------------------------------===//
+// MuxOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult MuxOp::fold(ArrayRef<Attribute> constants) {
+  // mux (a, b, b) -> b
+  if (trueValue() == falseValue())
+    return trueValue();
+
+  // mux(0, a, b) -> b
+  // mux(1, a, b) -> a
+  auto pred = constants[0].dyn_cast_or_null<IntegerAttr>();
+  if (pred) {
+    if (pred.getValue().isNullValue()) {
+      return falseValue();
+    } else {
+      return trueValue();
+    }
+  }
+
+  return {};
+}
+
+//===----------------------------------------------------------------------===//
+// ICmpOp
+//===----------------------------------------------------------------------===//
+
+// Calculate the result of a comparison when the LHS and RHS are both
+// constants.
+static bool applyCmpPredicate(ICmpPredicate predicate, const APInt &lhs,
+                              const APInt &rhs) {
+  switch (predicate) {
+  case ICmpPredicate::eq:
+    return lhs.eq(rhs);
+  case ICmpPredicate::ne:
+    return lhs.ne(rhs);
+  case ICmpPredicate::slt:
+    return lhs.slt(rhs);
+  case ICmpPredicate::sle:
+    return lhs.sle(rhs);
+  case ICmpPredicate::sgt:
+    return lhs.sgt(rhs);
+  case ICmpPredicate::sge:
+    return lhs.sge(rhs);
+  case ICmpPredicate::ult:
+    return lhs.ult(rhs);
+  case ICmpPredicate::ule:
+    return lhs.ule(rhs);
+  case ICmpPredicate::ugt:
+    return lhs.ugt(rhs);
+  case ICmpPredicate::uge:
+    return lhs.uge(rhs);
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+// Returns the result of applying the predicate when the LHS and RHS are the
+// exact same value.
+static bool applyCmpPredicateToEqualOperands(ICmpPredicate predicate) {
+  switch (predicate) {
+  case ICmpPredicate::eq:
+  case ICmpPredicate::sle:
+  case ICmpPredicate::sge:
+  case ICmpPredicate::ule:
+  case ICmpPredicate::uge:
+    return true;
+  case ICmpPredicate::ne:
+  case ICmpPredicate::slt:
+  case ICmpPredicate::sgt:
+  case ICmpPredicate::ult:
+  case ICmpPredicate::ugt:
+    return false;
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+// Try to apply the predicate to an unknown LHS but with a constant RHS.
+// This is accomplished by reasoning about RHS which are on the upper or
+// lower boundary of possible values.
+static IntegerAttr tryApplyCmpPredicateToRHS(ICmpPredicate predicate,
+                                             IntegerAttr &rhs) {
```

Comment:
You can pass rhs by value: `IntegerAttr &rhs` -> `IntegerAttr rhs`, it is just a pointer in size.

Reply:
deleted this function